### PR TITLE
In-memory Engine: implement range load

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Ref #16141

What's Changed:

```commit-message
implement range load
```
This PR implements range load. 
A range will experience three phases before providing read services:
1. pending range
2. pending range with snapshot
3. range with snapshot loaded

When deciding to load a range, say r1, r1 becomes pending range. 

In apply thread at some time, r1, coupled with the rocksdb snapshot, becomes pending range with snapshot, and a load task will be scheduled to a specific worker which will load data from the snapshot into the in-memory engine.  At the same time, any write for this range will be cached. 

When the snapshot load finished, r1 becomes range with snapshot loaded and at some time in the apply thread, it will consume the cached writes. Then, r1 becomes a normal range in the in-memory engine and load completes.

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/region_cache_memory_engine/src/background.rs
Comment (afeinberg) [inline components/region_cache_memory_engine/src/background.rs:88]:
```suggestion
    pub fn schedule_task(
```

File: components/region_cache_memory_engine/src/engine.rs
Comment (afeinberg) [inline components/region_cache_memory_engine/src/engine.rs:248]:
```suggestion
                .schedule_task(BackgroundTask::LoadTask)
```

**Review by afeinberg — CHANGES_REQUESTED** [review]
Small nit.

File: components/region_cache_memory_engine/src/background.rs
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/background.rs:173]: Minor: Can it be a more general Iterator (e.g. Box<dyn Iterator> maybe), if the caller does not need to use BTreeSet's specific method. It seems to me the caller does not care the type at all. It just needs to iterate all ranges.

File: components/region_cache_memory_engine/src/background.rs
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/background.rs:245]:
```suggestion
        core.mut_range_manager().clear_ranges_in_gc();
```

File: components/region_cache_memory_engine/src/range_manager.rs
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/range_manager.rs:96]:
```suggestion
    pub(crate) loading_snapshot_ranges: VecDeque<(CacheRange, Arc<RocksSnapshot>)>,
```

File: components/region_cache_memory_engine/src/range_manager.rs
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/range_manager.rs:97]:
```suggestion
    pub(crate) loading_cached_write_ranges: Vec<CacheRange>,
```

File: components/region_cache_memory_engine/src/range_manager.rs
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/range_manager.rs:99]:
```suggestion
    ranges_in_gc: BTreeSet<CacheRange>,
```

File: components/region_cache_memory_engine/src/range_manager.rs
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/range_manager.rs:256]:
```suggestion
    pub fn set_ranges_in_gc(&mut self, ranges_in_gc: BTreeSet<CacheRange>) {
```

File: components/region_cache_memory_engine/src/range_manager.rs
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/range_manager.rs:272]:
```suggestion
            return Err(LoadFailedReason::Evicted);
```

File: components/region_cache_memory_engine/src/range_manager.rs
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/range_manager.rs:269]:
```suggestion
            return Err(LoadFailedReason::InGc);
```

File: components/region_cache_memory_engine/src/write_batch.rs
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/write_batch.rs:148]: I think this should be refined with log(n) algorithm, instead of O(n) check. 
What's more, for every write and very key to go through this, it's not inefficient. I guess we may have to do it in region level (that is for each region's round in apply thread, this check is done only once)

File: components/region_cache_memory_engine/src/background.rs
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/background.rs:248]: maybe give a named struct to the 3 entries' tube?

File: components/region_cache_memory_engine/src/write_batch.rs
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/write_batch.rs:67]: for every entry to go through maybe_cached, it may be too expensive. Ideally it's done once per region.

File: components/region_cache_memory_engine/src/write_batch.rs
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/write_batch.rs:63]: write() would make apply threads wait with each other when calls write_opt. To me here it must be able to run parallelly. Otherwise this would make apply further unscalable.

File: components/region_cache_memory_engine/src/engine.rs
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/engine.rs:233]: It's not efficient to blindly call write lock regardless if there's work to do or not. 
Instead, we can quickly check  with read lock (ideally not lock at all but rust may not allow that) if there's any work to do. If there's work to do, acquire the write lock.

File: components/region_cache_memory_engine/src/write_batch.rs
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/write_batch.rs:148]: We will limit the total amount of ranges to load at the same time, say 5. So linear scan is fine.

> What's more, for every write and very key to go through this, it's not inefficient.
At least for now, it's not called for every write. It will only be checked at `write_to_db` just like checking whether to write a key in the memory engine.

File: components/region_cache_memory_engine/src/background.rs
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/background.rs]: good catch. Thanks.

File: components/region_cache_memory_engine/src/background.rs
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/background.rs]: good catch. Thanks.

File: components/region_cache_memory_engine/src/write_batch.rs
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/write_batch.rs:74]: maybe it can be further optimized by fast checking if there's any need to call maybe_cached by just checking ranges_loading_snapshot and ranges_loading_cached_write is empty or not. 
The check can be a help function of range_manager. 
It's faster than iterate the whole buffer in normal cases.

File: components/region_cache_memory_engine/src/range_manager.rs
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/range_manager.rs:89]: the comment should be updated
```suggestion
    // poped and pushed into `ranges_loading_snapshot`. Then the data in the snapshot
```

File: components/region_cache_memory_engine/src/background.rs
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/background.rs:178]: It seems to me ranges_in_gc is always same as ranges. Is there any reason to have it separately?

**Review by tonyxuqqi — COMMENTED** [review]
the rest LGTM

File: components/region_cache_memory_engine/src/write_batch.rs
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/write_batch.rs]: Good suggestion.

File: components/region_cache_memory_engine/src/background.rs
Comment (afeinberg) [inline components/region_cache_memory_engine/src/background.rs:326]:
```suggestion
                            Err(e) => {
                                 error!("creating rocksdb iterator failed"; "cf" => cf, "err" => %e);
                            }
```

File: components/region_cache_memory_engine/src/background.rs
Comment (afeinberg) [inline components/region_cache_memory_engine/src/background.rs:303]: Should we spawn this as an async task in a background worker remote like we do with full compaction (which also involves an expensive rocksdb operation): https://github.com/tikv/tikv/blob/66847e9c5ac701e73da1366938a7e2306cacc671/components/raftstore/src/store/worker/compact.rs#L371

I believe, the answer here is no, as it would not be safe to perform a gc concurrently with a load, and (unlike full compaction) we do not intend to pause this thread. However, just wanted to check. Thoughts @tonyxuqqi ?

File: components/region_cache_memory_engine/src/background.rs
Comment (afeinberg) [inline components/region_cache_memory_engine/src/background.rs:329]: is this only going to fail in exceptional cases?

File: components/region_cache_memory_engine/src/range_manager.rs
Comment (afeinberg) [inline components/region_cache_memory_engine/src/range_manager.rs:350]: since loading and gc are both sync and run in the same worker, can this happen? see my earlier comment in background.rs

File: components/region_cache_memory_engine/src/write_batch.rs
Comment (afeinberg) [inline components/region_cache_memory_engine/src/write_batch.rs:169]: nit: document that this is an out param.

**Review by afeinberg — COMMENTED** [review]
Looks mostly good, but see my comments about handling, and the need for load to perhaps run asynchronously if we do not want it to be blocked by GC tasks/block GC tasks.

File: components/region_cache_memory_engine/src/background.rs
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/background.rs:329]: Yes, it should not be failed.

File: components/region_cache_memory_engine/src/range_manager.rs
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/range_manager.rs:350]: Load is async. It first push the range in the pending_ranges in the RangeManager, then a LoadTask can be scheduled to the worker.

File: components/region_cache_memory_engine/src/background.rs
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/background.rs:303]: Only the load of the range completes will the range be GCed. Here is a background worker, we may not need it to be done in another worker.

Comment (afeinberg) [reply components/region_cache_memory_engine/src/background.rs:303]: yes, but there is only one thread here. since GC and load cannot run concurrently. see how I did this for full compaction. this can be done later (i can do it), just leave a TODO.

Comment (tonyxuqqi) [reply components/region_cache_memory_engine/src/background.rs:303]: Load itself will be quite expensive and it may block GC, which is critical for in-memory-engine's performance. So I think it's better to have a background task for the load.

Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/background.rs:325]: All right, it has been doen.

File: components/region_cache_memory_engine/src/background.rs
Comment (afeinberg) [reply components/region_cache_memory_engine/src/background.rs:168]: or ``impl Iterator<Item = CacheRange> `` if possible.

File: components/region_cache_memory_engine/src/background.rs
Comment (afeinberg) [inline components/region_cache_memory_engine/src/background.rs:855]: if you support passing a callback(`Box<dyn FnOnce(Vec<CacheRange>) + Send>`) to pending_load (and subsequently to `LoadTask`), you can avoid a sleep. e.g.,

```rust
let (cb, fut) = paired_future_callback::<Vec<CacheRange>>::();
engine.handle_pending_load(cb); 
...
let _ = block_on(f1);
```

File: tests/integrations/range_cache_engine.rs
Comment (afeinberg) [inline tests/integrations/range_cache_engine.rs:44]: this may be a good place to test load concurrent with gc.

**Review by afeinberg — APPROVED** [review]
Looking from write/apply path pov:

Looks good, but see comments from Tony about iterator as well as my comment about, re: adding a callback and testing load together with gc. I'll give this a provisional LGTM, so that once @tonyxuqqi approves and gives his own LGTM we can merge.

File: components/region_cache_memory_engine/src/range_manager.rs
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/range_manager.rs:350]: You mean for the same range, load and gc performs concurrently?

File: components/region_cache_memory_engine/src/background.rs
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/background.rs:269]: Good catch. Here I use read lock to filter the no write batch case and forget the adjust the code.

File: components/region_cache_memory_engine/src/range_manager.rs
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/range_manager.rs:350]: Actually, set_ranges_in_gc is to avoid the range being gced and loaded simutaneously.

File: components/region_cache_memory_engine/src/background.rs
Comment (LykxSassinator) [inline components/region_cache_memory_engine/src/background.rs:262]:
```suggestion
            core.has_cached_write_batch(&range)
```

File: components/region_cache_memory_engine/src/engine.rs
Comment (LykxSassinator) [inline components/region_cache_memory_engine/src/engine.rs:173]:
```suggestion
    pub(crate) fn has_cached_write_batch(&self, cache_range: &CacheRange) -> bool {
```

**Review by LykxSassinator — APPROVED** [review]

**Pull request merged**
